Fix multi-segment-type sliders getting mangled on legacy export

bdach opened the pull request:
- Closes https://github.com/ppy/osu/issues/31713
- Closes https://github.com/ppy/osu/issues/31714

[Previously](https://github.com/ppy/osu/pull/12303), [previously](https://github.com/ppy/osu/pull/12309), [previously](https://github.com/ppy/osu/pull/12310).

While I don't see immediate breakage in light testing, I accept that this may turn out to be entirely incorrect from the start line, so please treat with as much skepticism as you see warranted.

The logic in `LegacyBeatmapEncoder` that was supposed to handle the lazer-exclusive feature of supporting multiple slider segment types in a single slider was interfering rather badly with the Bezier converter. Generally it was a bit difficult to follow, too.

The nice thing about `BezierConverter` is that it is *guaranteed* to only output Bezier control points. In light of this, the same double-up-the-control-point logic that was supposed to make multiple slider segment types backwards-compatible with stable can be placed in the Bezier conversion logic, and be *much* more understandable, too. So I did that instead of attempting to jam more exceptions into the existing encoder code.

@OliBomby tagging in, as you may be interested in checking out this proposal, and maybe you can spot if I'm missing something in all this as well.

---

This change fixes some sliders which use mutliple curve types not getting correctly exported to the stable `.osz` beatmap package format.

bdach committed:
Move implicit slider path segment handling logic to Bezier converter

The logic in `LegacyBeatmapEncoder` that was supposed to handle
the lazer-exclusive feature of supporting multiple slider segment types
in a single slider was interfering rather badly with the Bezier
converter. Generally it was a bit difficult to follow, too.

The nice thing about `BezierConverter` is that it is *guaranteed* to
only output Bezier control points. In light of this, the same double-up-
-the-control-point logic that was supposed to make multiple slider
segment types backwards-compatible with stable can be placed in
the Bezier conversion logic, and be *much* more understandable, too.

OliBomby commented:
I'm not a fan of doing the double-up-the-control-point logic in `BezierConverter` because I designed that class to be used for lazer slider to lazer slider conversion, so the output of the conversion should still be compatible with lazer and not double-up the control points. I'd rather see the double-up-the-control-point logic being done in `LegacyBeatmapExporter`.

Removing the backwards compatibility logic from `LegacyBeatmapEncoder` and having it always do explicit segments seems fine to me.

bdach commented:
> because I designed that class to be used for lazer slider to lazer slider conversion, so the output of the conversion should still be compatible with lazer and not double-up the control points

Genuine question, do you see any future use for this method other than the current one?

I'm not opposed to moving the logic out as you propose, but it'll be a little annoying to write and I kinda wonder if it's worth the added code overhead.

OliBomby commented:
> Genuine question, do you see any future use for this method other than the current one?
> 
> I'm not opposed to moving the logic out as you propose, but it'll be a little annoying to write and I kinda wonder if it's worth the added code overhead.

One niche use-case I see is being able to do affine transformations on the shape of perfect curve type sliders. For example converting to bezier and then scaling it on the X axis to turn the circle into an ellipse.

Besides that I just think it will make the code less confusing, because all the backwards compatibility logic will be localized to `LegacyBeatmapExporter`.

bdach committed:
Move control point double-up logic to `LegacyBeatmapExporter`

Done for two reasons:

- During review it was requested for the logic to be moved out of
  `BezierConverter` as `BezierConverter` was intended to produce
  "lazer style" sliders with per-control-point curve types,
  as a future usability / code layering concern.

- It is also relevant for encode-decode stability. With how the logic
  was structured between the Bezier converter and the legacy beatmap
  encoder, the encoder would leave behind per-control-point Bezier curve
  specs that stable ignored, but subsequent encodes and decodes in lazer
  would end up multiplying the doubled-up control points ad nauseam.
  Instead, it is sufficient to only specify the curve type for the
  head control point as Bezier, not specify any further curve types
  later on, and instead just keep the double-up-control-point for new
  implicit segment logic which is enough to make stable cooperate
  (and also as close to outputting the slider exactly as stable would
  have produced it as we've ever been)

bdach commented:
Turns out I needed to do another thing to fix decode-encode stability that got broken with this, so it works out. Logic moved out in https://github.com/ppy/osu/pull/31734/commits/b4f63da048e16c9f0fd0d339ea13f33637dade9c.

OliBomby commented on osu.Game/Database/LegacyBeatmapExporter.cs:
This should not double the control point for `i == 0`. The first control point is always considered a new segment.

bdach committed:
Do not double up first control point of path

peppy reviewed (approved):
Fixes said issue and code reads better to me, win-win.
